### PR TITLE
Syshealth app: Add await-readiness stats and adapt to Scheduler

### PR DESCRIPTION
**Pull request by ianull**
I'm gonna generate latest performance report, so need to add new metrics and Scheduler adaption.

Changes:
* Adapt the StorageCore to single-threaded Scheduler
* Add a new metric to track awaitReady() latency which includes the overhead of the very first model synchronization as well as Scheduler things.

### CODE REVIEW

**Comment by ianull** [comment]
@yuangu-google sorry, I know the PR is kind of messing up your stability work. :(
But fwiw the adaption to new Scheduler arch should be required by both of stability and performance.

**Comment by ianull** [comment]
@jasonwyatt pls. help double check if I'm using handle.dispatcher correctly for fetch/store/close/etc. I'll base on the Scheduler adaption to generate new performance numbers.

File: javatests/arcs/android/systemhealth/testapp/StorageCore.kt [hunk 1]
```diff
@@ -278,73 +280,93 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
...
+                                TaskEvent(
+                                    TaskEventId.HANDLE_STORE_END,
+                                    System.currentTimeMillis(),
+                                    entity?.number
                                 )
-                            }
+                            )
                         }
                     }
                 }
```
Comment (jasonwyatt) [inline javatests/arcs/android/systemhealth/testapp/StorageCore.kt:320]: This part doesn't need to occur on the dispatcher, callback registration is thread-safe.

File: javatests/arcs/android/systemhealth/testapp/StorageCore.kt [hunk 2]
```diff
@@ -278,73 +280,93 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
...
+                    TestEntity.Companion
+                ),
+                when (settings.storageMode) {
+                    TestEntity.StorageMode.PERSISTENT -> TestEntity.singletonPersistentStorageKey
+                    else -> TestEntity.singletonInMemoryStorageKey
+                }
+            ) as ReadWriteSingletonHandle<TestEntity>
+
+            taskHandle.handle = withContext(handle.dispatcher) {
+                val elapsedTime = measureTimeMillis { handle.awaitReady() }
```
Comment (jasonwyatt) [inline javatests/arcs/android/systemhealth/testapp/StorageCore.kt:297]: this also doesn't need to be called within the dispatcher... awaitReady registers an onReady listener.

File: javatests/arcs/android/systemhealth/testapp/StorageCore.kt [hunk 3]
```diff
@@ -278,73 +280,93 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
...
+                                TaskEvent(
+                                    TaskEventId.HANDLE_STORE_END,
+                                    System.currentTimeMillis(),
+                                    entity.map { it.number }.toSet()
                                 )
-                            }
+                            )
                         }
                     }
                 }
```
Comment (jasonwyatt) [inline javatests/arcs/android/systemhealth/testapp/StorageCore.kt:362]: Same comments here...

File: javatests/arcs/android/systemhealth/testapp/StorageCore.kt [hunk 4]
```diff
@@ -397,10 +423,14 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
         when (val handle = taskHandle.handle) {
             is ReadWriteSingletonHandle<*> -> (
                 handle as? ReadWriteSingletonHandle<TestEntity>
-                                              )?.store(entity)?.join()
+                )?.let {
+                    withContext(it.dispatcher) { it.store(entity).join() }
```
Comment (jasonwyatt) [inline javatests/arcs/android/systemhealth/testapp/StorageCore.kt:427]: Not a good idea to join within the dispatcher context.. better to put `join` after the closing brace.

File: javatests/arcs/android/systemhealth/testapp/StorageCore.kt [hunk 5]
```diff
@@ -397,10 +423,14 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
...
+                }
             is ReadWriteCollectionHandle<*> -> (
                 handle as? ReadWriteCollectionHandle<TestEntity>
-                                               )?.store(entity)?.join()
+                )?.let {
+                    withContext(it.dispatcher) { it.store(entity).join() }
```
Comment (jasonwyatt) [inline javatests/arcs/android/systemhealth/testapp/StorageCore.kt:432]: Same

**Pull request merged**